flow/enhancers/snort.go

```diff
@@ -0,0 +1,200 @@
+
+const (
+	timestamp = iota
+	snortRAW
+	decodeHEX
+	END
+)
+
+const sepHEX = "- -   - - - - - - - - - - - -  - - - - - - - - - - - -  - - - - - - - - -\n"
+
+func min(a, b int) int {
+	if a < b {
+		return a
+	}
+	return b
+}
+
+// SnortFlowEnhancer describes a snort graph enhancer
+type SnortFlowEnhancer struct {
+	cache *cache.Cache
+}
+
+// Enhance A and B node TID based on neutron database
+func (sfe *SnortFlowEnhancer) Enhance(f *flow.Flow) {
+	if s, ok := sfe.cache.Get(f.UUID); ok {
+		msgs := s.([]*flow.SnortMessage)
+		f.Snort = append(f.Snort, msgs...)
+		sfe.cache.Delete(f.UUID)
+	}
+}
+
+func flowFromSnortMessage(msg *flow.SnortMessage) *flow.Flow {
+	parentUUID := ""
+	L2ID := int64(0)
+	L3ID := int64(0)
+	nodeTID := ""
+	gpkt := gopacket.NewPacket(msg.Data, layers.LayerTypeEthernet, gopacket.NoCopy)
+	new := flow.NewFlow()
+	key := flow.KeyFromGoPacket(&gpkt, parentUUID).String()
+
+	new.Init(key, msg.Timestamp, &gpkt, int64(len(msg.Data)), nodeTID, parentUUID, L2ID, L3ID)
+	return new
+}
+
+func parseSnortTimestamp(timestamp string) int64 {
+	t := time.Now()
+	var months, days, hour, min, sec, usec int
+	fmt.Sscanf(timestamp, "%d/%d-%d:%d:%d.%d",
+		&months, &days, &hour, &min, &sec, &usec)
+	year, _, _ := t.Date()
+	return time.Date(year, time.Month(months), days, hour, min, sec, usec*1000, t.Location()).UTC().Unix()
+}
+
+func (sfe *SnortFlowEnhancer) recvSnortMessage(msg *flow.SnortMessage) {
+	var msgs []*flow.SnortMessage
+	f := flowFromSnortMessage(msg)
+	if s, ok := sfe.cache.Get(f.UUID); ok {
+		cacheMsgs := s.([]*flow.SnortMessage)
+		msgs = append(msgs, cacheMsgs...)
+	}
+	msgs = append(msgs, msg)
+	sfe.cache.Set(f.UUID, msgs, cache.DefaultExpiration)
+}
+
+// readCMGX : read snort log with running with "-A cmg -X" options
+func (sfe *SnortFlowEnhancer) parseSnortCMGX(reader *bufio.Reader) error {
+	regexTimestamp := regexp.MustCompile("^(?P<timestamp>\\d\\d/\\d\\d-\\d\\d:\\d\\d:\\d\\d\\.\\d\\d\\d\\d\\d\\d) \\[.+?\\] \\[.+?\\] \\\"(?P<message>.+?)\\\" \\[.+?\\] \\[(?P<classification>.+?)\\].*\n")
+	regexSnortRAW := regexp.MustCompile("^snort\\.raw\\[(?P<bytes>\\d+)\\]:\n")
+
+	var packetBytes int
+	var msg *flow.SnortMessage
+	step := timestamp
+	for {
+		str, err := reader.ReadString('\n')
+		if err != nil {
+			return err
+		}
+		if step == timestamp && regexTimestamp.MatchString(str) {
+			msg = &flow.SnortMessage{}
+			msg.Timestamp = parseSnortTimestamp(regexTimestamp.ReplaceAllString(str, "${timestamp}"))
+			msg.Message = regexTimestamp.ReplaceAllString(str, "${message}")
+			msg.Classification = regexTimestamp.ReplaceAllString(str, "${classification}")
+			step = snortRAW
+			continue
+		}
+		if step == snortRAW && regexSnortRAW.MatchString(str) {
+			packetBytes, err = strconv.Atoi(regexSnortRAW.ReplaceAllString(str, "${bytes}"))
+			if err != nil {
+				return err
+			}
+			step = decodeHEX
+			continue
+		}
+		if step == decodeHEX {
```

Comment:
Wouldn't it open skydive to a potential dos ? For example, since the connection is not authenticated, a attacker could just connect, and feed skydive with a crafted message of unlimited hexadecimal line, resulting in a growing msg.Data until there is OOM and skydive get killed.

Reply:
technically more or less yes, but take in account the cache will force timeout the entires that didn't match for amount of time.